File: ios/bridge/ExpoBridgePlayer.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgePlayer {
  static func setPlaybackQueue(queueService: QueueService, itemId: String, type: String) async throws -> String {
    try await ExpoBridge.asyncBridge {
      try await queueService.setQueue(itemId: itemId, type: type)
      return "Track(s) added to queue"
    }
  }

  static func getCurrentState(
    playbackController: PlaybackController
  ) async -> [String: Any] {
    let state = playbackController.state
    let songInfo = await playbackController.fetchCurrentSongInfo()
    return BridgeResponses.playbackState(
      playbackRate: state.playbackRate,
      playbackStatus: MusicItemMapper.describePlaybackStatus(state.playbackStatus),
      playbackTime: playbackController.playbackTime,
      currentSong: songInfo
    )
  }

  static func playLibrarySong(
    queueService: QueueService,
    musicUserToken: String,
    songId: String
  ) async throws -> String {
    try await ExpoBridge.asyncBridge {
      try await queueService.playLibrarySong(musicUserToken: musicUserToken, songId: songId)
      return "Library song added to queue"
    }
  }

  static func playLibraryPlaylist(
    queueService: QueueService,
    musicUserToken: String,
    playlistId: String,
    startingAt: Int
  ) async throws -> String {
    try await ExpoBridge.asyncBridge {
      try await queueService.playLibraryPlaylist(
        musicUserToken: musicUserToken,
        playlistId: playlistId,
        startingAt: startingAt
      )
      return "Library playlist added to queue"
    }
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgePlayer.kt
package expo.modules.applemusic.bridge

import com.apple.android.music.playback.model.MediaPlayerException
import expo.modules.applemusic.AndroidPlaybackController
import expo.modules.applemusic.AndroidQueueService
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

internal fun ModuleDefinitionBuilder.registerPlayerBridge(
  moduleScope: CoroutineScope,
  playbackController: () -> AndroidPlaybackController,
  queueService: () -> AndroidQueueService,
  emitPlaybackError: (Exception, String) -> Unit,
  emitPlaybackTimeUpdate: (Double) -> Unit,
) {
  AsyncFunction("setPlaybackQueue") Coroutine { itemId: String, type: String ->
    try {
      queueService().setQueue(itemId, type)
      "Track(s) added to queue"
    } catch (error: Exception) {
      throw AndroidPlaybackController.mapPlaybackException(error)
    }
  }

  AsyncFunction("playLibrarySong") Coroutine { musicUserToken: String, songId: String ->
    queueService().playLibrarySong(musicUserToken, songId)
    "Library song added to queue"
  }

  AsyncFunction("playLibraryPlaylist") Coroutine { musicUserToken: String, playlistId: String, startingAt: Int ->
    queueService().playLibraryPlaylist(musicUserToken, playlistId, startingAt)
    "Library playlist added to queue"
  }

  AsyncFunction("configurePlayer") { options: Map<String, Any?> ->
    playbackController().configurePlayer(options)
  }

  AsyncFunction("getCurrentState") {
    playbackController().currentState()
  }

  Function("play") {
    runPlayback(moduleScope, emitPlaybackError, "play") { playbackController().play() }
  }

  Function("pause") {
    playbackController().pause()
  }

  Function("skipToNextEntry") {
    runPlayback(moduleScope, emitPlaybackError, "skipToNext") { playbackController().skipToNext() }
  }

  Function("skipToPreviousEntry") {
    runPlayback(moduleScope, emitPlaybackError, "skipToPrevious") { playbackController().skipToPrevious() }
  }

  Function("restartCurrentEntry") {
    playbackController().restartCurrentEntry { time -> emitPlaybackTimeUpdate(time) }
  }

  Function("seekToTime") { time: Double ->
    playbackController().seekToTime(time) { actual -> emitPlaybackTimeUpdate(actual) }
  }

  Function("togglePlayerState") {
    runPlayback(moduleScope, emitPlaybackError, "togglePlayback") { playbackController().togglePlayback() }
  }
}

private fun runPlayback(
  scope: CoroutineScope,
  emitPlaybackError: (Exception, String) -> Unit,
  operation: String,
  block: () -> Unit,
) {
  scope.launch {
    try {
      block()
    } catch (error: MediaPlayerException) {
      emitPlaybackError(error, operation)
    } catch (error: Exception) {
      emitPlaybackError(error, operation)
    }
  }
}
